ENIGMAsystem/SHELL/Graphics_Systems/General/GSmodel.cpp

```diff
@@ -227,27 +236,44 @@ void d3d_model_primitive_begin(int id, int kind, int format) {
 
 void d3d_model_primitive_end(int id) {
   enigma::Model* model = enigma::models[id];
+  #ifdef DEBUG_MODE
+  if (!model->current_primitive) {
+    show_error("d3d_model_primitive_end called for model " + std::to_string(id) + " and no current primitive exists!\n" \
+               "This can occur if you call end without actually calling begin.", false);
+    return;
+  }
+  #endif
   // copy current primitive to the stack
   enigma::Primitive primitive = *model->current_primitive;
   // delete the current primitive from the heap
   delete model->current_primitive;
   model->current_primitive = 0;
 
-  if (model->use_draw_color && !model->vertex_colored) {
-    vertex_color(model->vertex_buffer, model->vertex_color, model->vertex_alpha);
-  }
-
-  // if we are guessing the format of the model, end the vertex format now
-  if (!vertex_format_exists(primitive.format)) {
+  // if the primitive doesn't have a valid vertex format
+  // and one does exist that we were guessing, then end
+  // that vertex format now
+  if (!vertex_format_exists(primitive.format) && vertex_format_exists()) {
     if (!primitive.format_defined && model->use_draw_color && !model->vertex_colored) {
       vertex_format_add_color();
     }
-    model->vertex_colored = false;
     primitive.format = vertex_format_end();
     primitive.format_defined = true;
   }
 
+  // if the last vertex didn't specify a color then we have to do so now
+  if (model->use_draw_color && !model->vertex_colored) {
+    vertex_color(model->vertex_buffer, model->vertex_color, model->vertex_alpha);
+    model->vertex_colored = true;
+  }
+
   size_t vertex_size = vertex_get_buffer_size(model->vertex_buffer) - primitive.vertex_offset;
+  // don't keep empty primitives
+  if (!vertex_size) {
+    #ifdef DEBUG_MODE
+    show_error("A primitive was ended that had 0 size on model: " + std::to_string(id), false);
```

Comment:
I'm not sure if this warrants an error, either, but I guess it can be until (A) we have better warning-reporting logic and/or (B) someone complains.

Reply:
I tested, you are correct that it's NOT an error, even in debug mode, in GM8.1 or GMSv1.4.
```
mymodel = d3d_model_create();
d3d_model_primitive_begin(mymodel, pr_trianglelist);
d3d_model_primitive_end(mymodel);
```

Reply:
I tweaked it now: e738a07ec392ab20cd8ba7435f2ccda060ae9c1e

This will only occur now if you specify a format that exists during primitive begin and then don't specify vertex data. In run mode the size will not be checked and the primitive will be kept.